kwasm-lemmas.md

```diff
@@ -382,22 +367,20 @@ They are non-trivial in their implementation, but the following should obviously
       [simplification]
 
     rule #getRange(ByteMap <| .Map |>, _, _) => 0 [simplification]
-```
 
-`#get` over `#setRange`.
-
-```k
-    rule #get(#setRange(BM, SET_ADDR, VAL, WIDTH), GET_ADDR) => #wrap(1, VAL)
-        requires SET_ADDR ==Int GET_ADDR
-         andBool #isByteMap(BM)
-        [simplification]
-
-    rule #get(#setRange(BM, SET_ADDR       , VAL        , WIDTH       ), GET_ADDR)
-      => #get(#setRange(BM, SET_ADDR +Int 1, VAL >>Int 8, WIDTH -Int 1), GET_ADDR)
-        requires #isByteMap(BM)
-         andBool GET_ADDR >Int SET_ADDR
-         andBool GET_ADDR <Int SET_ADDR +Int WIDTH
-        [simplification]
+    rule #getRange(#setRange(BM, ADDR, VAL, WIDTH), ADDR', WIDTH') => #getRange(BM, ADDR', WIDTH') requires ADDR' +Int WIDTH' <=Int ADDR  [simplification]
+    rule #getRange(#setRange(BM, ADDR, VAL, WIDTH), ADDR', WIDTH') => #getRange(BM, ADDR', WIDTH') requires ADDR  +Int WIDTH  <=Int ADDR' [simplification]
+    rule #getRange(#setRange(BM, ADDR, VAL, WIDTH), ADDR', WIDTH') => VAL
+      requires 0 <=Int ADDR
+       andBool 0  <Int WIDTH
+       andBool 0 <=Int VAL andBool VAL <Int 2 ^Int (8 *Int WIDTH)
+       andBool ADDR'  ==Int ADDR
+       andBool WIDTH' ==Int WIDTH
+       [simplification]
+
+    syntax Int ::= #byteWidth ( Int ) [function]
+ // --------------------------------------------
+    rule #byteWidth(#getRange(_, _, N)) => N
```

Comment:
This is not well-defined as a function, because `#getRange(.Map, _, W) == 0`, but `#byteWidth(#getRange(.Map, _, W)) == W`, where W can be integer. Is the idea to have `#getRange` actually return bytes later? Otherwise we should change this. You could encode it as a relationship instead, like

```
   syntax Int ::= #byteWidthLessThan(Int, Int) [function]
   rule #byteWidthLessThan(#getRange(_, _, N), M) => true requires N <=Int M
```
and possibly more cases for other `Int`s if necessary.

But looking at how it's used, you might be better off just matching on the `#getRange` in the simplifications where you use this function. Your call.

Reply:
A relation may be possible, but I don't need this to be a total function, I just need it to work on a handful of cases where it's used. I'll see if I can refactor it, but it does simplify one of the rules a bit.

Reply:
This is not really a matter of it being total but rather of it not being a function at all, since not every input maps to a single output. It's not true that `a = b ==> f(a) = f(b)`. For example, `#getRange(.Map, _, 2)) = #getRange(.Map, _, 3))`, but `#byteWidth(#getRange(.Map, _, 2)) =/= #byteWidth(#getRange(.Map, _, 3))`

Reply:
Hmmm, good point. I'll remove it.